Bump cardano-node to 1.35.4

Anviking opened the pull request:
- [x] Somehow use [CHaP](https://github.com/input-output-hk/cardano-haskell-packages) to get the right revisions
- [x] Keep `nix-shell` working 
- [x] Are we actually using the correct ledger revisions?
    - Seems right https://github.com/input-output-hk/plutus/blame/d5f6f4a5505094490f61c8d164515adf7c8f46cc/plutus-ledger-api/src/Plutus/ApiCommon.hs#L147
- [x] Get `cabal build` working
    - [x] Fix warnings of the type `src/Cardano/Pool/DB/Sqlite/TH.hs:56:1: warning: [-Wname-shadowing]`
- [x] De-duplicate source of truth for compiler version and index state
- [x] Use non-broken version of `io-classes` to fix integration tests
    - Can't use 0.2.0.0 / f4183f274d88d0ad15817c7052df3a6a8b40e6dc
- [x] Misc cleanup
- [x] Update readme matrix


### Comments

<!-- Additional comments, links, or screenshots to attach, if any. -->

### Issue Number

ADP-2382

jonathanknowles commented on lib/text-class/text-class.cabal:
```diff
@@ -35,7 +35,6 @@ library
     , text
     , time
     , hspec
-    , OddWord
```
What's the issue with the `OddWord` package?

Anviking replied:
It believe it broke with GHC 9.2. However I realised cardano-node, unlike many of its dependencies, isn't quite supporting GHC 9.2 yet anyway, so it's still 8.10.7 we should be using.

jonathanknowles replied:
Fair enough!

I've added a task to the PR description to "fix OddWord breakage". (Just to remind ourselves.)

Anviking replied:
Already reverted 😉

Anviking commented on cabal.project:
```diff
@@ -38,7 +38,23 @@
+    c00aae8461a256275598500ea0e187588c35a5d5d7454fb57eac18d9edb86a56
+    d4a35cd3121aa00d18544bb0ac01c3e1691d618f462c46129271bccf39f7e8ee
+
+-- See CONTRIBUTING for information about these, including some Nix commands
```
Note was copied from cardano-node, so need to either remove this note or also adjust CONTRIBUTING. Probably we'd prefer to update something in the `docs/contributing` directory.

Anviking commented on cabal.project:
```diff
@@ -56,36 +72,7 @@ packages:
-    location: https://github.com/haskell-works/hw-aeson
-    tag: ba7c1e41c6e54d6bf9fd1cd013489ac713fc3153
-    --sha256: 1czyn0whgv7czzgwn5y1zgwlkb9p9sn9z9sc9gixrjprcyc0p15p
+  flags: -support_rdrand -integer-gmp
```
Not sure the `-integer-gmp` was actually needed in the end 🤔 
```suggestion
  flags: -support_rdrand
```

Anviking commented on cabal.project:
```diff
@@ -133,152 +108,22 @@ source-repository-package
+    tag: a91add165152fa36f08e95fafe7da24f1dba4690
+    --sha256: 1ia8vlqghis92cla8qmqa6kh8f3jn29b01fshyk5hmgy5373s684
 
+-- FIXME: Is this revision too new?
```
```suggestion
```

Anviking committed:
Make cabalProjectIndexState not break with CHaP

When the second line is added to cabal.project:
```
index-state: 2022-09-27T00:00:00Z
index-state: cardano-haskell-packages 2022-10-25T20:00:00Z
```

we still need to be able to extract the index state.

Anviking committed:
Bump to cardano-node 1.35.4 & start using CHaP

Anviking committed:
Bump ProtocolVersion in integration cluster

Anviking commented on lib/wallet/api/http/Cardano/Wallet/Launch/Cluster.hs:
```diff
@@ -915,7 +915,7 @@ generateGenesis dir systemStart initialFunds addPoolsToGenesis = do
             -- the protocol version rather than just the era, so we need to
             -- set it to a realisitic value.
-            , _protocolVersion = Ledger.ProtVer 7 0
+            , _protocolVersion = Ledger.ProtVer 8 0
```
Will enable https://github.com/input-output-hk/plutus/blame/d5f6f4a5505094490f61c8d164515adf7c8f46cc/plutus-ledger-api/src/Plutus/ApiCommon.hs#L147, though not that we have anything in our integration tests that would use it...

Anviking committed:
update node compatibility matrix in readme

michalrus reviewed (changes requested):
Hey, we (Daedalus / Lace Desktop) are seeing differing hashes between Linux and Darwin `nix-prefetch-url` of https://github.com/input-output-hk/cardano-wallet/archive/c8c9cd96e94bd06ec2ce75620e17055161724003.tar.gz

Further investigation in [this thread](https://input-output-rnd.slack.com/archives/CG1FBSDMM/p1669636993754929) led to finding that case (in)sensitivity in paths is to blame. When you run a diff of unpacked TAR archive between Linux and Darwin, you see:

```diff
% diff -u <(cat tree-linux | sort) <(ssh mw@10.77.2.237 cat tree-darwin | sort)
--- /proc/self/fd/11	2022-11-28 14:10:28.709470308 +0100
+++ /proc/self/fd/12	2022-11-28 14:10:28.709470308 +0100
@@ -308,16 +308,15 @@
 ./lib/primitive/test/Cardano/Wallet/Primitive/Types/UTxOSelectionSpec.hs
 ./lib/primitive/test/Cardano/Wallet/Primitive/Types/UTxOSelectionSpec/TypeErrorSpec.hs
 ./lib/primitive/test/Cardano/Wallet/Primitive/Types/UTxOSpec.hs
-./lib/primitive/test/data
 ./lib/primitive/test/Data
-./lib/primitive/test/data/Cardano
-./lib/primitive/test/data/Cardano/Wallet
-./lib/primitive/test/data/Cardano/Wallet/Primitive
-./lib/primitive/test/data/Cardano/Wallet/Primitive/Types
-./lib/primitive/test/data/Cardano/Wallet/Primitive/Types/TokenMap
-./lib/primitive/test/data/Cardano/Wallet/Primitive/Types/TokenMap/FlatTokenMap.json
-./lib/primitive/test/data/Cardano/Wallet/Primitive/Types/TokenMap/NestedTokenMap.json
-./lib/primitive/test/data/Cardano/Wallet/Primitive/Types/TokenQuantity.json
+./lib/primitive/test/Data/Cardano
+./lib/primitive/test/Data/Cardano/Wallet
+./lib/primitive/test/Data/Cardano/Wallet/Primitive
+./lib/primitive/test/Data/Cardano/Wallet/Primitive/Types
+./lib/primitive/test/Data/Cardano/Wallet/Primitive/Types/TokenMap
+./lib/primitive/test/Data/Cardano/Wallet/Primitive/Types/TokenMap/FlatTokenMap.json
+./lib/primitive/test/Data/Cardano/Wallet/Primitive/Types/TokenMap/NestedTokenMap.json
+./lib/primitive/test/Data/Cardano/Wallet/Primitive/Types/TokenQuantity.json
 ./lib/primitive/test/Data/QuantitySpec.hs
 ./lib/primitive/test/run-test-suite.hs
 ./lib/primitive/test/Spec.hs
```

So you can see that Linux contains both `data` *and* `Data`. This is worth fixing, since checking out on Windows or Darwin might do weird things.

Unisay commented on cabal.project:
```diff
@@ -38,7 +38,23 @@
+-- Custom repository for cardano haskell packages, see
+-- https://github.com/input-output-hk/cardano-haskell-packages
+-- for more information.
+repository cardano-haskell-packages
```
YaY! ✌🏼

Unisay commented on cabal.project:
```diff
@@ -291,44 +135,44 @@ source-repository-package
 
+-- Temporary patch for version 0.1.0.0 (181601bc3d9e9d21a671ce01e0b481348b3ca104)
+-- in order to change io-sim from 606de33fa2f467d108fb1efb86daeb3348bf34e3
+-- to 57e888b1894829056cb00b7b5785fdf6a74c3271 to include the ReaderT fix.
```
Lets make sure we don't forget to remove this temp patch by creating a JIRA task and linking it here.

Anviking replied:
Again, this seems like a place where simply leaving a comment designed to be discovered exactly when it is needed seems useful. I also see less upsides here, at least if creating a task specific to this alone; this can't be tackled without the next node bump...

...but I could create a task for the next node bump and link/comment. The only problem is that we don't necessarily know for sure what version (and task title) it will be, but that seems fine anyway.

Anviking replied:
https://input-output.atlassian.net/browse/ADP-2447

Unisay commented on nix/overlays/common-lib.nix:
```diff
@@ -39,13 +39,24 @@ in {
 
     # Get the index-state and ghc version out of a cabal.project file.
+    #
+    # TODO: It would be better to re-use functions from haskell.nix like:
```
Lets add a JIRA & tag it with "devops" and reference here.

Anviking replied:
Created and added reference to https://input-output.atlassian.net/browse/ADP-2445

Unisay commented on cabal.project:
```diff
@@ -367,12 +210,48 @@ constraints:
+  , HSOpenSSL >= 0.11.7.2
+  , algebraic-graphs < 0.7
+  , protolude < 0.3.1
+    -- TODO: these should be set in cabal files, but avoiding setting them in lower dependencies for initial CHaP release
```
Lets create another JIRA task and reference it here.
BTW, I don't quite understand the explanation in the comment.

Anviking replied:
🙈 This is a comment copied from cardano-node in the `  -- Copied from cardano-node's cabal.project:` section — I guess I can remove this `TODO` in it.

Anviking commented on cabal.project:
```diff
@@ -367,12 +210,48 @@ constraints:
+  , protolude < 0.3.1
+    -- TODO: these should be set in cabal files, but avoiding setting them in lower dependencies for initial CHaP release
+  , cardano-prelude == 0.1.0.0
+--  , cardano-prelude <0.1.0.1
```
```suggestion
```

Anviking committed:
fixups

Anviking committed:
Link TODO to ADP-2445

Anviking committed:
Add TODOs for next node bump

Unisay reviewed (approved):
🚢 👍🏼

michalrus reviewed (approved):
> 👎 Rejected by code reviews

Sorry, accepted!

Anviking committed:
Add `cliEraFlag` to keep cluster setup working in alonzo

jonathanknowles commented on cabal.project:
```diff
@@ -291,44 +135,48 @@ source-repository-package
-  location: https://github.com/input-output-hk/typed-protocols
-  tag: 181601bc3d9e9d21a671ce01e0b481348b3ca104
-  --sha256: 1lr97b2z7l0rpsmmz92rsv27qzd5vavz10cf7n25svya4kkiysp5
+  location: https://github.com/Anviking/typed-protocols.git
```
Hi @Anviking!

Just curious as to the reason for depending on this fork of `typed-protocols` here?
GitHub claims that:
```
This branch is up to date with input-output-hk/typed-protocols:main. 
```
Were you planning to add more features to it? 😃

Anviking replied:
It's explained just above. It's for updating a dependency (`io-sim`), and should no longer be needed with the next node bump.

I had preferred using a branch in the iohk repo, but I don't have push access, and as it will go away next bump, I think this is fine.

Anviking replied:
I was also surprised that a `source-repository-package` for `io-sim` wasn't enough. _Double-checking the typed-protocols fork can't be dropped..._

=> Seems like I was able to drop it 🤨 I'm confused why it seemed to fail before, but that's good.

Anviking committed:
Drop unnecessary dependency on typed-protocols fork
